server/resp/src/main/java/org/infinispan/server/resp/RespRequestHandler.java

```diff
@@ -1,16 +1,109 @@
+    * @param arguments The remaining arguments to the command
+    * @return stage that when complete returns the new handler to instate. This stage <b>must</b> be completed on the event loop
+    */
+   public CompletionStage<RespRequestHandler> handleRequest(ChannelHandlerContext ctx, String type, List<byte[]> arguments) {
+      if ("QUIT".equals(type)) {
+         ctx.close();
+         return myStage;
+      }
       ctx.writeAndFlush(stringToByteBuf("-ERR unknown command\r\n", ctx.alloc()));
-      return this;
+      return myStage;
+   }
+
+   protected <E> CompletionStage<RespRequestHandler> stageToReturn(CompletionStage<E> stage, ChannelHandlerContext ctx,
+         BiConsumer<? super E, Throwable> biConsumer) {
+      return stageToReturn(stage, ctx, Objects.requireNonNull(biConsumer), null);
+   }
+
+   protected <E> CompletionStage<RespRequestHandler> stageToReturn(CompletionStage<E> stage, ChannelHandlerContext ctx,
+         Function<E, RespRequestHandler> handlerWhenComplete) {
+      return stageToReturn(stage, ctx, null, Objects.requireNonNull(handlerWhenComplete));
+   }
+
+   /**
+    * Handles ensuring that a stage BiConsumer is only invoked on the event loop. The BiConsumer can then do things
+    * such as writing to the underlying channel or update variables in a thread safe manner without additional
+    * synchronization or volatile variables.
+    * <p>
+    * If the BiConsumer provided is null the provided stage <b>must</b> be completed only on the event loop of the
+    * provided ctx.
+    * This can be done via the Async methods on {@link CompletionStage} such as {@link CompletionStage#thenApplyAsync(Function, Executor)};
+    * <p>
+    * This method can only be invoked on the event loop for the provided ctx
+    *
+    * @param <E> The stage return value
+    * @param stage The stage that will complete. Note that if biConsumer is null this stage may only be completed on the event loop
+    * @param ctx The context used for this request, normally the event loop is the thing used
+    * @param biConsumer The consumer to be ran on
+    * @param handlerWhenComplete A function to map the result to which handler will be used for future requests. Only ever invoked on the event loop. May be null, if so the current handler is returned
+    * @return A stage that is only ever completed on the event loop that provides the new handler to use
+    */
+   private <E> CompletionStage<RespRequestHandler> stageToReturn(CompletionStage<E> stage, ChannelHandlerContext ctx, BiConsumer<? super E, Throwable> biConsumer,
+         Function<E, RespRequestHandler> handlerWhenComplete) {
+      assert ctx.channel().eventLoop().inEventLoop();
+      if (CompletionStages.isCompletedSuccessfully(stage)) {
+         E result;
+         try {
+            result = CompletionStages.join(stage);
+            if (biConsumer != null) {
+               biConsumer.accept(result, null);
+            }
+         } catch (Throwable t) {
```

Comment:
In this code path, `biConsumer.accept` is invoked twice if `biConsumer.accept(result, null);` throws an exception.
I believe you want 2 try-catch blocks:
* one around `join`
* and another around `biConsumer.accept`

Reply:
Yeah, the assumption was that biConsumer would not be throwing an exception. I can handle that case as you mention.

Reply:
Reworked this to be cleaner and more defines how the exception handling works